server/metrics/interceptors.go

```diff
@@ -0,0 +1,38 @@
+package metrics
+
+import (
+	"context"
+	"path/filepath"
+	"time"
+
+	"google.golang.org/grpc"
+)
+
+// UnaryInterceptor adds all necessary metrics to incoming gRPC requests
+func UnaryInterceptor() grpc.UnaryServerInterceptor {
+	return func(
+		ctx context.Context,
+		req interface{},
+		info *grpc.UnaryServerInfo,
+		handler grpc.UnaryHandler,
+	) (interface{}, error) {
+		// start values
+		operationStart := time.Now()
+		operation := filepath.Base(info.FullMethod)
+
+		// the RPC
+		resp, err := handler(ctx, req)
+
+		// record the operation
+		CRIOOperations.WithLabelValues(operation).Inc()
+		CRIOOperationsLatency.WithLabelValues(operation).
+			Observe(SinceInMicroseconds(operationStart))
+
+		// record error metric if occurred
+		if err != nil {
+			CRIOOperationsErrors.WithLabelValues(operation).Inc()
+		}
+
+		return resp, err
+	}
+}
```

Comment:
Maybe it's pre-tea time, but how/when does this get called from say image_status.go?  Is it part of the grpc processing that you get for free?

Reply:
Yes, the interceptors here are registered there as part of the gRPC request processing:
https://github.com/cri-o/cri-o/pull/2809/files#diff-702431da3a2e35392cf253955d15f774R741-R744

This means this function will be called for every request, like we do now for the logging of the requests and responses in debug mode.